### ISSUE TITLE
Can we instantiate the model outside the "run" block?

### ISSUE THREAD

**Issue by AdityaSoni19031997** [pytorch/xla#1819]
## ❓ Questions and Help

As shown in the example [colab nbs](https://github.com/pytorch/xla/blob/master/contrib/colab/multi-core-alexnet-fashion-mnist.ipynb), The model named "net" was defined within the "run" function, What if i do it outside the run function? IS it the correct way? If not, What's the difference b/w the approaches?
(When using Multiple TPU Cores)

Sample Code (inspired from the notebook linked)

```python
def run():
    ....
    ....
    net = model()
    net = net.to(device)
    ....
    ....
xmp.spawn(map_fn, args=(flags,), nprocs=8, start_method='fork')

Vs

net_global = model()
def run():
    ....
    ....
    # net = Model()
    net = net_global.to(device)
    ....
    .....
xmp.spawn(map_fn, args=(flags,), nprocs=8, start_method='fork')

Vs

net = model()
def run():
    ....
    ....
    # net = Model()
    net.to(device)
    ....
    .....
xmp.spawn(map_fn, args=(flags,), nprocs=8, start_method='fork')
```

Secondly,
```python
  # Downloads train and test datasets
  # Note: each process has its own (identical) copies of each dataset.
  #  This is for simplicity and is not the only supported pattern.
  train_dataset = datasets.FashionMNIST(
    dataset_path,
    train=True,
    download=True,
    transform=my_transform)
```

From the colab notebook linked above, as per last line, there are other ways as well? Does that mean that we can define datasets/dataloaders outside the "run" as well? If not can little light be thrown on "This is for simplicity and is not the only supported pattern."?

Thanks.

**Comment by dlibenzi** [comment]
It is not advised to create data structures outside the MP function target call stack.
That does not mean you have to create them directly inside the function, as long as they are created from any function which has the MP target function as parent in the call stack.
Or simply, avoid creating source code where models are global (that is, they get created when the module is loaded, without any function to be called).
When using the pytorch multiprocessing `spawn` start_method, any global data will have to be pickled, so that puts further restrictions on what you can create on global scope.

While for models that **might** work, you cannot call any XLA function besides `xmp.spawn()` (the set is a bit wider, but it is better to stay on the safe side) outside of the MP target function call stack.

Same issue with pickling applies to datasets.
On top of that, for multicore training you will need distributed samplers, which need the core ordinal, which is only available once you are within the MP target function call stack.

**Comment by AdityaSoni19031997** [comment]
Thanks Davide for your reply. Just curious about this,

When we use the model variable which is defined publicly Vs the model
variable which is defined with the MP function, we observe that the code
crashes due to OOM in the later case but not in the former approach.

Would you please comment little more wrt what's causing that memory spike
and ultimately an OOM?

Just want to understand more on how pytorch-xla works/differs from using TF
for TPUs. Pointers would be highly appreciated.

Thanks a lot for your help!

**Comment by dlibenzi** [comment]
OK, so, if you have:
```Python
import torch
import torch_xla
...
BIG_CPU_TENSOR = build_pytorch_cpu_tensor()
...
```
**AND** you are using `start_method=fork`, then the memory pages which are backing up `BIG_CPU_TENSOR` will be COW-ed (marked Copy On Write) by `fork(2)` and unless you are writing on them, you can save memory.

But we advise to use `start_method=fork` only on Colab because pytorch seems not happy with it (this might be mostly a pytorch GPU limitation, but I am not 100% sure).
With `start_method=spawn` that `BIG_CPU_TENSOR` will be pickled and sent to the process, so there won't be any COW advantage.

**Comment by AdityaSoni19031997** [comment]
Thanks Again!
Here's another question, I cannot promise it's the last one from me but :smile:

>TF is directly creating the replicas on the TPU, while PyTorch XLA makes the replicas on the VM, then distributes them to the TPU. 

Is this what is happening?

**Comment by dlibenzi** [comment]
This is pytorch, not TF, and it follows the same semantics.
Doing multi-processing GPU in pytorch is exactly the same. One process per GPU.

While it could be possible to change the execution model, that would probably deviate the API from normal pytorch patterns, which is not what we aimed for.
And it will also become useless once we migrate the Cloud TPU architecture to use a single VM (which would unleash more cores and RAM on the client side).
